### PR TITLE
nixos/forgejo: refactor secrets, add `cfg.secrets`, forgejo: build `environment-to-ini`, nixos/tests/forgejo: test `cfg.secrets` using /metrics endpoint

### PR DESCRIPTION
**Pull request by emilylange**
## Description of changes

Please refer to the commit messages for details.
I see very little reason to copy them word for word into this PR description :)

More or less announced in https://github.com/NixOS/nixpkgs/issues/262802#issuecomment-1777159444.

I've been using `environment-to-ini` on NixOS long before Forgejo even existed.
It did require some fairly hacky overrides and involved manually escaping the environment variables and then loading them from a single `serviceConfig.EnvironmentFile`.

Finally found the time and spoons to do it properly and upstream it.

IMHO this improves things A LOT.

Open to lots of nitpicking, reformatting and so on :p

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by soopyc — APPROVED** [review]

**Comment by soopyc** [comment]
I think this is notable enough to warrant an entry in the release notes. 

Might also be a good idea to set the defaults in `options` rather than `config`, or at least document them so users don't have to dig through the source to see them.

**Comment by emilylange** [comment]
> I think this is notable enough to warrant an entry in the release notes.
> 
> Might also be a good idea to set the defaults in `options` rather than `config`, or at least document them so users don't have to dig through the source to see them.

I would prefer to do that in a follow-up PR instead.

I have a proof of concept in my worktree locally that moves the automatically bootstrapped secrets from `config` to `options`, making them easier to discover, as requested.
It also adds another `secretsBootstrap` (boolean) option to disable the bootstrapping entirely.

That diff is, however, relatively big and somewhat opinionated (in the way that it tries to define all values once, instead of repeating them all over the place, at the expense of making it more difficult to follow, at first).

I expect it to take multiple review rounds, and we may even opt to use a completely different implementation (e.g. that repeats itself) altogether.

As such, I would like to get this PR here reviewed and merged first, before opening another one which further improves this.
Just so the diff doesn't get too big and the changes to controversial, which would make this PR unnecessarily difficult and depending on who you ask impossible to review.

Would love to hear your release note entry suggestion though, if you happen to have anything at hand right now :)
If you don't, that's fine, I'll try to come up with one over the next few days.
I agree that this should probably be part of the release notes, and we should probably do that in this PR here, instead of the follow-up one.

File: nixos/modules/services/misc/forgejo.nix [hunk 1]
```diff
@@ -501,61 +584,15 @@ in
...
-              ${optionalString (cfg.mailerPasswordFile != null) ''
-                ${replaceSecretBin} '#mailerpass#' '${cfg.mailerPasswordFile}' '${runConfig}'
-              ''}
-              chmod u-w '${runConfig}'
+              config='${cfg.customDir}/conf/app.ini'
+              cp -f '${format.generate "app.ini" cfg.settings}' "$config"
+
+              chmod u+w "$config"
+              ${lib.getExe' cfg.package "environment-to-ini"} --config "$config"
+              chmod u-w "$config"
```
Comment (steveej) [inline nixos/modules/services/misc/forgejo.nix:595]: this leads to the following error for me:

```
gitea[612]: 2024/06/02 18:04:21 ...es/setting/oauth2.go:151:loadOAuth2From() [F] save oauth2.JWT_SECRET failed: failed to save "/var/lib/forgejo/custom/conf/app.ini": open /var/lib/forgejo/custom/conf/app.ini: permission denied
```

i'm using this changed module with an existing forgejo installation data directory, _and_ i've changed the `oauth2.JWT_SECRET` from what was autogenerated.

Comment (emilylange) [reply nixos/modules/services/misc/forgejo.nix:595]: This looks to me like you messed up something in your config.

From what I can tell, Forgejo only does that when both `oauth2.JWT_SECRET` and `oauth2.JWT_SECRET_URI` are missing or are invalid (e.g. wrong format, wrong length).

Can you please share your NixOS config, more than just a single line of logs, and, if you can, either your secret or how you generated it?

Comment (steveej) [reply nixos/modules/services/misc/forgejo.nix:595]: i can happily confirm my messing up :smile: 

the invalid token was generated using `openssl rand -hex 32` (command i found randomly online) and forgejo generated a new one as it wasn't suitable. i copied the new one into my secrets file now as a quick fix. eventually i'd like to look into how to pre-generate these secrets.

it didn't require a config change to fix my mistake. here it is anyway:

```nix    
    sops.secrets.FORGEJO_JWT_SECRET = {};
    sops.secrets.FORGEJO_INTERNAL_TOKEN = {};
    sops.secrets.FORGEJO_SECRET_KEY = {};

    services.forgejo = {
      enable = true;
      package = repoFlake.inputs.nixpkgs_forgejo.legacyPackages.${pkgs.system}.forgejo;
      settings = {
        service.DISABLE_REGISTRATION = true;
        server.HTTP_ADDR = "127.0.0.1";
        server.DISABLE_SSH = true;
        server.ROOT_URL = "https://forgejo.${domain}";
        server.HTTP_PORT = 3001;
      };
      secrets = {
        oauth2.JWT_SECRET = lib.mkForce config.sops.secrets.FORGEJO_JWT_SECRET.path;
        security.INTERNAL_TOKEN = lib.mkForce config.sops.secrets.FORGEJO_INTERNAL_TOKEN.path;
        security.SECRET_KEY = lib.mkForce config.sops.secrets.FORGEJO_SECRET_KEY.path;
      };
    };
```

**Review by steveej — APPROVED** [review]
thank you :pray: this works great and is a big improvement :+1:

**Review by bendlas — APPROVED** [review]
Neat!

**Review by adamcstephens — APPROVED** [review]

**Review by pyrox0 — APPROVED** [review]
LGTM, thank you for contributing this, it's very helpful!

**Comment by emilylange** [comment]
Had to rebase for the `24.11` release note entry that was requested by @soopyc in https://github.com/NixOS/nixpkgs/pull/312523#issuecomment-2118855532.

3 pushes in total happened:

1. Rebase to the HEAD of `master` at the time (**force push**)
1. `nixos/rl-2405: add services.forgejo.secrets` (note how it says `2405` despite being for `2411`)
1. Commit message typo fixup (`nixos/rl-2405: add services.forgejo.secrets` -> `nixos/rl-241: add services.forgejo.secrets`) (**force push**)

**Review by adamcstephens — APPROVED** [review]

**Review by soopyc — APPROVED** [review]

**Pull request merged**